### PR TITLE
ZMNHVD1 sensor settings

### PR DESCRIPTION
**Pull request by hfaas**
Hi Athom,

I've extended the settings for the ZMNHVD1 in a an attemt to fix the problems of reading I1 when having a sensor attached.

The attemt of fixing the reading issues failed, but I think the settings might be a good additional feature set for the application.

I hope you guys can soon fix the I1 sensor input reading (correctly representing in the app and making sure it also correctly sets the output, although the last should theoretically be possible with the PID loop of the device).

I'm not sure the naming of the settings templates is well chosen, but that is up to you. ;)

Kind regards

### CODE REVIEW

File: .homeycompose/drivers/settings/I1autoManualControl.json [hunk 1]
```diff
@@ -0,0 +1,33 @@
+{
+  "id": "autoManualControl",
+  "type": "dropdown",
+  "label": {
+    "en": "Auto or manual control",
+    "nl": "Automatische of manuele sturing"
```
Comment (RobinBol) [inline .homeycompose/drivers/settings/I1autoManualControl.json]: The label of this setting should reflect what it changes, not the available options. For example: label: 'Control type' and values: 'Manual' | 'Auto'.

File: .homeycompose/drivers/settings/I1autoManualControl.json [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    "index": 52,
+    "size": 1
+  },
+  "value": "0",
+  "values": [
+    {
+      "id": "0",
+      "label": {
+        "en": "Manual/disabled",
+        "nl": "Manueel/disabled"
```
Comment (RobinBol) [inline .homeycompose/drivers/settings/I1autoManualControl.json]: 'handmatige' sound better imo

File: .homeycompose/drivers/settings/I1autoManualControl.json [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    {
+      "id": "1",
+      "label": {
+        "en": "Automatic",
+        "nl": "Automatisch"
+      }
+    }
+  ],
+  "hint": {
+    "en": "This parameter ({{zwaveParameterIndex}}) determines how the input should invluence the output (manual=diabled or automatic=output follows input using PID).\n\nNote: This parameter only invluences the output when the input type parameter (1) is configured for a 0-10V sensor (values 3, 4 and 5).",
```
Comment (RobinBol) [inline .homeycompose/drivers/settings/I1autoManualControl.json]: influence*
disabled*

It is not really clear what you are trying to explain here (manual=dia...using PID).
It is not really clear what you are trying to explain here (This parameter... values 3, 4 and 5). How are these values related to this setting?

File: .homeycompose/drivers/settings/I1autoManualControl.json [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    "nl": "Deze parameter ({{zwaveParameterIndex}}) bepaalt hoe de ingang de outgang beinvloed (manueel=niet of automatisch=uitgang volgt ingang middels PID).\n\nLet op: Deze parameter beinvloed de uitgang alleen wanneer de ingang type parameter (1) is ingesteld op een 0-10V sensor (waardes 3, 4 of 5)."
```
Comment (RobinBol) [inline .homeycompose/drivers/settings/I1autoManualControl.json]: uitgang*
beïnvloed*

It is not really clear what you are trying to explain here (manueel=niet... middels PID).
It is not really clear what you are trying to explain here (Deze parameter... waarde 3, 4 of 5). How are these values related to this setting?

File: .homeycompose/drivers/settings/I1maximumSensorValue.json [hunk 5]
```diff
@@ -0,0 +1,24 @@
+{
+  "type": "number",
+  "label": {
+    "en": "Maximum sensor value (°C/Lux/%)",
```
Comment (RobinBol) [inline .homeycompose/drivers/settings/I1maximumSensorValue.json]: No units in the label, especially if there are defined in `units`.

File: .homeycompose/drivers/settings/I1maximumSensorValue.json [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    "nl": "Maximale sensor niveau (°C/Lux/%)"
+  },
+  "units": {
+    "en": "",
```
Comment (RobinBol) [inline .homeycompose/drivers/settings/I1maximumSensorValue.json]: Missing `unit.en` property.

Comment (RobinBol) [reply .homeycompose/drivers/settings/I1maximumSensorValue.json]: Are these units correct? Are these the only possible types?

File: .homeycompose/drivers/settings/I1minimumSensorValue.json [hunk 7]
```diff
@@ -0,0 +1,24 @@
+{
+  "type": "number",
```
Comment (RobinBol) [inline .homeycompose/drivers/settings/I1minimumSensorValue.json]: See comments with `I1maximumSensorValue.json`.

File: .homeycompose/drivers/settings/I1sensorTriggerThreshold.json [hunk 8]
```diff
@@ -0,0 +1,24 @@
+{
+  "type": "number",
+  "label": {
+    "en": "I1 sensor trigger threshold",
```
Comment (RobinBol) [inline .homeycompose/drivers/settings/I1sensorTriggerThreshold.json]: We don't use references like 'I1', use 'Input 1' instead.

File: drivers/ZMNHVD1/driver.settings.compose.json [hunk 9]
```diff
@@ -74,6 +74,9 @@
       },
       {
         "$extends": "doubleClick"
+      },
+      {
+        "$extends": "I1autoManualControl"
```
Comment (RobinBol) [inline drivers/ZMNHVD1/driver.settings.compose.json]: We don't use references like 'I1', use 'Input 1' instead.

File: drivers/ZMNHVD1/driver.settings.compose.json [hunk 10]
```diff
@@ -119,6 +122,27 @@
       }
     ]
   },
+  {
+    "type": "group",
+    "label": {
+      "en": "I1 input sensor reporting",
```
Comment (RobinBol) [inline drivers/ZMNHVD1/driver.settings.compose.json]: We don't use references like 'I1', use 'Input 1' instead.

File: drivers/ZMNHVD1/driver.settings.compose.json [hunk 11]
```diff
@@ -119,6 +122,27 @@
...
+      "nl": "I1 ingang sensor metingen"
+    },
+    "children": [
+      {
+        "$extends": "I1minimumSensorValue"
```
Comment (RobinBol) [inline drivers/ZMNHVD1/driver.settings.compose.json]: We don't use references like 'I1', use 'Input 1' instead.

File: drivers/ZMNHVD1/driver.settings.compose.json [hunk 12]
```diff
@@ -119,6 +122,27 @@
...
+      },
+      {
+        "$extends": "I1maximumSensorValue"
```
Comment (RobinBol) [inline drivers/ZMNHVD1/driver.settings.compose.json]: We don't use references like 'I1', use 'Input 1' instead.

File: drivers/ZMNHVD1/driver.settings.compose.json [hunk 13]
```diff
@@ -119,6 +122,27 @@
...
+      },
+      {
+        "$extends": "I1sensorReportingThreshold"
```
Comment (RobinBol) [inline drivers/ZMNHVD1/driver.settings.compose.json]: We don't use references like 'I1', use 'Input 1' instead.

File: drivers/ZMNHVD1/driver.settings.compose.json [hunk 14]
```diff
@@ -119,6 +122,27 @@
...
+      },
+      {
+        "$extends": "I1sensorTriggerThreshold"
```
Comment (RobinBol) [inline drivers/ZMNHVD1/driver.settings.compose.json]: We don't use references like 'I1', use 'Input 1' instead.

File: .homeycompose/drivers/settings/I1sensorReportingThreshold.json [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+    "size": 2
+  },
+  "value": 5,
+  "attr": {
+    "min": 0,
+    "max": 10000
+  },
+  "hint": {
+    "en": "This parameter ({{zwaveParameterIndex}}) defines the relative threshold which when exceeded will result in a sensor value report being send to Homey. Note: setting this parameter to 0 disables sensor reports completely.\n\nNote: This parameter only applies when the input type parameter (1) is configured for a 0-10V sensor (values 3, 4 and 5).",
+    "nl": "Deze parameter ({{zwaveParameterIndex}}) bepaalt de relatieve verandering die overschreden moet worden voordat een sensor waarde report wordt verstuurd naar Homey. Let op: het instellen van deze parameter op 0 schakelt sensor reports volledig uit.\n\nLet op: Deze parameter wordt alleen gebruikt wanneer de ingang type parameter (1) is ingesteld op een 0-10V sensor (waardes 3, 4 of 5)."
```
Comment (RobinBol) [inline .homeycompose/drivers/settings/I1sensorReportingThreshold.json]: 'sensor updates' instead of 'sensor reports'.

File: .homeycompose/drivers/settings/I1sensorTriggerThreshold.json [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+    "nl": "I1 sensor trigger drempelwaarde"
```
Comment (RobinBol) [inline .homeycompose/drivers/settings/I1sensorTriggerThreshold.json]: 'sensor update drempelwaarde'

File: .homeycompose/drivers/settings/I1sensorTriggerThreshold.json [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+    "index": 141,
+    "size": 1
+  },
+  "value": 5,
+  "attr": {
+    "min": 1,
+    "max": 100
+  },
+  "hint": {
+    "en": "This parameter ({{zwaveParameterIndex}}) defines the fixed threshold (percentage) which when exceeded will result in a basic set on association group 2.\n\nNote: This parameter only applies when the input type parameter (1) is configured for a 0-10V sensor (values 3, 4 and 5).",
```
Comment (RobinBol) [inline .homeycompose/drivers/settings/I1sensorTriggerThreshold.json]: not sure if 'fixed' adds anything here

File: .homeycompose/drivers/settings/I1sensorTriggerThreshold.json [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+    "nl": "Deze parameter ({{zwaveParameterIndex}}) bepaalt de vaste drempel (in procenten) die overschreden moet worden voordat een basic set bericht op associatie group 2 gepost wordt.\n\nLet op: Deze parameter wordt alleen gebruikt wanneer de ingang type parameter (1) is ingesteld op een 0-10V sensor (waardes 3, 4 of 5)."
```
Comment (RobinBol) [inline .homeycompose/drivers/settings/I1sensorTriggerThreshold.json]: Not sure if 'vaste' adds anything here

Comment (RobinBol) [reply .homeycompose/drivers/settings/I1sensorTriggerThreshold.json]: 'voordat een basic set wordt verstuurd naar associatiegroep 2'

File: .homeycompose/drivers/settings/I1autoManualControl.json (hunk 3 above)
Comment (hfaas) [reply .homeycompose/drivers/settings/I1autoManualControl.json]: I understand your comment, but this is a direct reference to the device manual. I'm not sure how better to explain this setting. This setting defines whether the input controls the out (using PID).
Also if parameter 1 (input type) is not set to one of the sensor types (value 3,4 of 5) this parameter is ignored by the device (e.g. in case it is set to a switch input). See also the device [manual](https://qubino.com/manuals/Flush_Dimmer_0-10V.pdf).

File: .homeycompose/drivers/settings/I1autoManualControl.json (hunk 4 above)
Comment (hfaas) [reply .homeycompose/drivers/settings/I1autoManualControl.json]: I understand your comment, but this is a direct reference to the device manual. I'm not sure how better to explain this setting. This setting defines whether the input controls the out (using PID).
Also if parameter 1 (input type) is not set to one of the sensor types (value 3,4 of 5) this parameter is ignored by the device (e.g. in case it is set to a switch input). See also the device [manual](https://qubino.com/manuals/Flush_Dimmer_0-10V.pdf).

File: .homeycompose/drivers/settings/I1maximumSensorValue.json (hunk 5 above)
Comment (hfaas) [reply .homeycompose/drivers/settings/I1maximumSensorValue.json]: How to deal with multiple units? Unit is sensor type dependent.

File: drivers/ZMNHVD1/driver.settings.compose.json (hunk 11 above)
Comment (hfaas) [reply drivers/ZMNHVD1/driver.settings.compose.json]: These are "file references" do you really want Input1 (without the white space of course) as prefix (also all following comments)?

File: .homeycompose/drivers/settings/I1sensorTriggerThreshold.json [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+    "size": 1
+  },
+  "value": 5,
+  "attr": {
+    "min": 1,
+    "max": 100
+  },
+  "hint": {
+    "en": "This parameter ({{zwaveParameterIndex}}) defines the threshold (percentage) which when exceeded will result in a basic set on association group 2.\n\nNote: This parameter only applies when the input type parameter (1) is configured for a 0-10V sensor (values 3, 4 and 5).",
+    "nl": "Deze parameter ({{zwaveParameterIndex}}) bepaalt de drempel (in procenten) die overschreden moet worden voordat een basic set wordt verstuurd naar associatiegroup 2.\n\nLet op: Deze parameter wordt alleen gebruikt wanneer de ingang type parameter (1) is ingesteld op een 0-10V sensor (waardes 3, 4 of 5)."
```
Comment (RobinBol) [inline .homeycompose/drivers/settings/I1sensorTriggerThreshold.json]: associatiegroep*

File: drivers/ZMNHVD1/driver.settings.compose.json (hunk 11 above)
Comment (RobinBol) [reply drivers/ZMNHVD1/driver.settings.compose.json]: Yes, this way it is consistent with the other input setting templates. Sorry if I'm nitpicking, but consistency is important:)

File: drivers/ZMNHVD1/driver.settings.compose.json [hunk 20]
```diff
@@ -119,6 +122,27 @@
...
+        "$extends": "Input1maximumSensorValue"
+      },
+      {
+        "$extends": "Input1sensorReportingThreshold"
+      },
+      {
+        "$extends": "Input1sensorTriggerThreshold"
+      }
+    ]
+  },
```
Comment (RobinBol) [inline drivers/ZMNHVD1/driver.settings.compose.json]: I think it is better to add these input settings to the input settings group, and not a separate group.

**Comment by RobinBol** [comment]
Thanks for the improvements @hfaas . I added some textual changes myself, would you take a look at the remaining comments? I think we can merge it afterwards.

File: drivers/ZMNHVD1/driver.settings.compose.json (hunk 20 above)
Comment (hfaas) [reply drivers/ZMNHVD1/driver.settings.compose.json]: I've re-worked all naming to be in line with the other settings.

**Comment by RobinBol** [comment]
@hfaas when you think your PR is ready for another review please re-request a review with me! (I saw not all comments were fixed yet, so I assume you are still working on it).

File: drivers/ZMNHVD1/driver.settings.compose.json (hunk 11 above)
Comment (hfaas) [reply drivers/ZMNHVD1/driver.settings.compose.json]: Don't worry about the nitpicking, I can relate to that. I totally agree about being consistent is important. 
I've completely renamed all items to <type><function><input>.json naming (as most others are). So in this case "sensorMinimumValueInput1.json".

**Comment by hfaas** [comment]
Hi @RobinBol 
Thanks for your reply. I've been going over your comments again, but except for the  comment about the hint information (of which I replied it is a direct reference to the manual and I'm not sure how to explain it better), I believe I've covered all your comments.

Which comments did I mis?

File: drivers/ZMNHVD1/driver.settings.compose.json (hunk 20 above)
Comment (RobinBol) [reply drivers/ZMNHVD1/driver.settings.compose.json]: This one :) I think it would make more sense to have these 4 input related settings being moved to the 'Inputs/Ingangen' settings group. What do you think?

Comment (hfaas) [reply drivers/ZMNHVD1/driver.settings.compose.json]: Aah, sorry I couldn't respond earlier. 
I mis understood (read it wrongly), and thought you were talking about the naming!
The reason I did it like this, was that this had a similar setup as the temperature therefore it is debatable, but I do understand what you mean. I would say whatever you think fits best.

Comment (hfaas) [reply drivers/ZMNHVD1/driver.settings.compose.json]: I see you already merged it into master (probably due to my late response), do you still want me to change this?

Comment (RobinBol) [reply drivers/ZMNHVD1/driver.settings.compose.json]: Ah, thanks for the clarification. I had a release scheduled yesterday so I was so free to merge it and change it myself so it could be released in one go (install it from https://homey.app/a/com.qubino/test/). Forgot to mention it to you yesterday.
